sphinx 5 compatibility

In one of MNE-Python's CIs we're using sphinx 5.0 beta, and seeing this deprecation warning percolating up from docutils:

```
Traceback (most recent call last):
  File "/home/circleci/python_env/lib/python3.8/site-packages/sphinx/events.py", line 94, in emit
    results.append(listener.handler(self.app, *args))
  File "/home/circleci/python_env/lib/python3.8/site-packages/numpydoc/numpydoc.py", line 104, in relabel_references
    for citation_node in doc.traverse(citation):
  File "/home/circleci/python_env/lib/python3.8/site-packages/docutils/nodes.py", line 240, in traverse
    warnings.warn('nodes.Node.traverse() is obsoleted by Node.findall().',
PendingDeprecationWarning: nodes.Node.traverse() is obsoleted by Node.findall().
```

I can probably take a swing at this in the next week or so (?) but would be happy if someone else picked it up too

We are getting the following failures with `docutils == 0.18.1` in  #402:

```
=================================== FAILURES ===================================
_________________________ test_reference[html_file0-1] _________________________

sphinx_app = <sphinx.application.Sphinx object at 0x7f9274a6a5b0>
html_file = ['index.html'], expected_length = 1

    @pytest.mark.parametrize(
        ("html_file", "expected_length"),
        (
            (["index.html"], 1),
            (["generated", "numpydoc_test_module.my_function.html"], 1),
            (["generated", "numpydoc_test_module.MyClass.html"], 1),
        ),
    )
    def test_reference(sphinx_app, html_file, expected_length):
        """Test for bad references"""
        out_dir = sphinx_app.outdir
    
        with open(op.join(out_dir, *html_file)) as fid:
            html = fid.read()
    
        reference_list = re.findall(r'<a class="fn-backref" href="\#id\d+">(.*)<\/a>', html)
    
>       assert len(reference_list) == expected_length
E       assert 0 == 1
E        +  where 0 = len([])

expected_length = 1
fid        = <_io.TextIOWrapper name='/tmp/pytest-of-runner/pytest-0/root/_build/html/index.html' mode='r' encoding='UTF-8'>
html       = '\n<!DOCTYPE html>\n\n<html lang="en">\n  <head>\n    <meta charset="utf-8" />\n    <meta name="viewport" content="wid...ref="_sources/index.rst.txt"\n          rel="nofollow">Page source</a>\n    </div>\n\n    \n\n    \n  </body>\n</html>'
html_file  = ['index.html']
out_dir    = '/tmp/pytest-of-runner/pytest-0/root/_build/html'
reference_list = []
sphinx_app = <sphinx.application.Sphinx object at 0x7f9274a6a5b0>

numpydoc/tests/test_full.py:94: AssertionError
_________________________ test_reference[html_file1-1] _________________________

sphinx_app = <sphinx.application.Sphinx object at 0x7f9274a6a5b0>
html_file = ['generated', 'numpydoc_test_module.my_function.html']
expected_length = 1

    @pytest.mark.parametrize(
        ("html_file", "expected_length"),
        (
            (["index.html"], 1),
            (["generated", "numpydoc_test_module.my_function.html"], 1),
            (["generated", "numpydoc_test_module.MyClass.html"], 1),
        ),
    )
    def test_reference(sphinx_app, html_file, expected_length):
        """Test for bad references"""
        out_dir = sphinx_app.outdir
    
        with open(op.join(out_dir, *html_file)) as fid:
            html = fid.read()
    
        reference_list = re.findall(r'<a class="fn-backref" href="\#id\d+">(.*)<\/a>', html)
    
>       assert len(reference_list) == expected_length
E       assert 0 == 1
E        +  where 0 = len([])

expected_length = 1
fid        = <_io.TextIOWrapper name='/tmp/pytest-of-runner/pytest-0/root/_build/html/generated/numpydoc_test_module.my_function.html' mode='r' encoding='UTF-8'>
html       = '\n<!DOCTYPE html>\n\n<html lang="en">\n  <head>\n    <meta charset="utf-8" />\n    <meta name="viewport" content="wid..._module.my_function.rst.txt"\n          rel="nofollow">Page source</a>\n    </div>\n\n    \n\n    \n  </body>\n</html>'
html_file  = ['generated', 'numpydoc_test_module.my_function.html']
out_dir    = '/tmp/pytest-of-runner/pytest-0/root/_build/html'
reference_list = []
sphinx_app = <sphinx.application.Sphinx object at 0x7f9274a6a5b0>

numpydoc/tests/test_full.py:94: AssertionError
_________________________ test_reference[html_file2-1] _________________________

sphinx_app = <sphinx.application.Sphinx object at 0x7f9274a6a5b0>
html_file = ['generated', 'numpydoc_test_module.MyClass.html']
expected_length = 1

    @pytest.mark.parametrize(
        ("html_file", "expected_length"),
        (
            (["index.html"], 1),
            (["generated", "numpydoc_test_module.my_function.html"], 1),
            (["generated", "numpydoc_test_module.MyClass.html"], 1),
        ),
    )
    def test_reference(sphinx_app, html_file, expected_length):
        """Test for bad references"""
        out_dir = sphinx_app.outdir
    
        with open(op.join(out_dir, *html_file)) as fid:
            html = fid.read()
    
        reference_list = re.findall(r'<a class="fn-backref" href="\#id\d+">(.*)<\/a>', html)
    
>       assert len(reference_list) == expected_length
E       assert 0 == 1
E        +  where 0 = len([])

expected_length = 1
fid        = <_io.TextIOWrapper name='/tmp/pytest-of-runner/pytest-0/root/_build/html/generated/numpydoc_test_module.MyClass.html' mode='r' encoding='UTF-8'>
html       = '\n<!DOCTYPE html>\n\n<html lang="en">\n  <head>\n    <meta charset="utf-8" />\n    <meta name="viewport" content="wid...test_module.MyClass.rst.txt"\n          rel="nofollow">Page source</a>\n    </div>\n\n    \n\n    \n  </body>\n</html>'
html_file  = ['generated', 'numpydoc_test_module.MyClass.html']
out_dir    = '/tmp/pytest-of-runner/pytest-0/root/_build/html'
reference_list = []
sphinx_app = <sphinx.application.Sphinx object at 0x7f9274a6a5b0>

numpydoc/tests/test_full.py:94: AssertionError
=============================== warnings summary ===============================
numpydoc/tests/test_full.py::test_MyClass
numpydoc/tests/test_full.py::test_MyClass
numpydoc/tests/test_full.py::test_MyClass
  /home/runner/work/numpydoc/numpydoc/numpydoc/numpydoc.py:104: PendingDeprecationWarning: nodes.Node.traverse() is obsoleted by Node.findall().
    for citation_node in doc.traverse(citation):

numpydoc/tests/test_full.py::test_MyClass
numpydoc/tests/test_full.py::test_MyClass
numpydoc/tests/test_full.py::test_MyClass
  /home/runner/work/numpydoc/numpydoc/numpydoc/numpydoc.py:84: PendingDeprecationWarning: nodes.Node.traverse() is obsoleted by Node.findall().
    section_node.traverse(

numpydoc/tests/test_full.py::test_MyClass
numpydoc/tests/test_full.py::test_MyClass
numpydoc/tests/test_full.py::test_MyClass
numpydoc/tests/test_full.py::test_MyClass
numpydoc/tests/test_full.py::test_MyClass
  /home/runner/work/numpydoc/numpydoc/numpydoc/numpydoc.py:124: PendingDeprecationWarning: nodes.Node.traverse() is obsoleted by Node.findall().
    for xref_node in ref.parent.traverse(matching_pending_xref):

numpydoc/tests/test_full.py::test_MyClass
numpydoc/tests/test_full.py::test_MyClass
numpydoc/tests/test_full.py::test_MyClass
  /home/runner/work/numpydoc/numpydoc/numpydoc/numpydoc.py:132: PendingDeprecationWarning: nodes.Node.traverse() is obsoleted by Node.findall().
    for ref in doc.traverse(reference, descend=True):

numpydoc/tests/test_full.py::test_MyClass
numpydoc/tests/test_full.py::test_MyClass
numpydoc/tests/test_full.py::test_MyClass
  /home/runner/work/numpydoc/numpydoc/numpydoc/numpydoc.py:135: PendingDeprecationWarning: nodes.Node.traverse() is obsoleted by Node.findall().
    for citation_node in doc.traverse(citation, descend=True):

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
-- generated xml file: /home/runner/work/numpydoc/numpydoc/junit-results.xml ---


=========================== short test summary info ============================
XFAIL numpydoc/tests/test_validate.py::TestValidator::test_bad_docstrings[BadParameters-blank_lines-msgs30]
XFAIL numpydoc/tests/test_validate.py::TestValidator::test_bad_docstrings[BadReturns-no_type-msgs33]
FAILED numpydoc/tests/test_full.py::test_reference[html_file0-1] - assert 0 == 1
FAILED numpydoc/tests/test_full.py::test_reference[html_file1-1] - assert 0 == 1
FAILED numpydoc/tests/test_full.py::test_reference[html_file2-1] - assert 0 == 1
============ 3 failed, 234 passed, 2 xfailed, 17 warnings in 4.27s =============
Error: Process completed with exit code 1.
```

--- a/numpydoc/numpydoc.py
+++ b/numpydoc/numpydoc.py
@@ -43,6 +43,15 @@ logger = logging.getLogger(__name__)
 HASH_LEN = 12
 
 
+def _traverse_or_findall(node, condition, **kwargs):
+    """Triage node.traverse (docutils <0.18) vs node.findall."""
+    return (
+        node.findall(condition, **kwargs)
+        if hasattr(node, "findall")
+        else node.traverse(condition, **kwargs)
+    )
+
+
 def rename_references(app, what, name, obj, options, lines):
     # decorate reference numbers so that there are no duplicates
     # these are later undecorated in the doctree, in relabel_references
@@ -81,8 +90,12 @@ def _is_cite_in_numpydoc_docstring(citation_node):
             return False
 
     sibling_sections = itertools.chain(
-        section_node.traverse(
-            is_docstring_section, include_self=True, descend=False, siblings=True
+        _traverse_or_findall(
+            section_node,
+            is_docstring_section,
+            include_self=True,
+            descend=False,
+            siblings=True,
         )
     )
     for sibling_section in sibling_sections:
@@ -101,7 +114,7 @@ def _is_cite_in_numpydoc_docstring(citation_node):
 
 def relabel_references(app, doc):
     # Change 'hash-ref' to 'ref' in label text
-    for citation_node in doc.traverse(citation):
+    for citation_node in _traverse_or_findall(doc, citation):
         if not _is_cite_in_numpydoc_docstring(citation_node):
             continue
         label_node = citation_node[0]
@@ -121,7 +134,7 @@ def relabel_references(app, doc):
                     and node[0].astext() == f"[{ref_text}]"
                 )
 
-            for xref_node in ref.parent.traverse(matching_pending_xref):
+            for xref_node in _traverse_or_findall(ref.parent, matching_pending_xref):
                 xref_node.replace(xref_node[0], Text(f"[{new_text}]"))
             ref.replace(ref_text, new_text.copy())
 
@@ -129,10 +142,10 @@ def relabel_references(app, doc):
 def clean_backrefs(app, doc, docname):
     # only::latex directive has resulted in citation backrefs without reference
     known_ref_ids = set()
-    for ref in doc.traverse(reference, descend=True):
+    for ref in _traverse_or_findall(doc, reference, descend=True):
         for id_ in ref["ids"]:
             known_ref_ids.add(id_)
-    for citation_node in doc.traverse(citation, descend=True):
+    for citation_node in _traverse_or_findall(doc, citation, descend=True):
         # remove backrefs to non-existent refs
         citation_node["backrefs"] = [
             id_ for id_ in citation_node["backrefs"] if id_ in known_ref_ids

--- a/numpydoc/numpydoc.py
+++ b/numpydoc/numpydoc.py
@@ -44,7 +44,11 @@ HASH_LEN = 12
 
 
 def _traverse_or_findall(node, condition, **kwargs):
-    """Triage node.traverse (docutils <0.18) vs node.findall."""
+    """Triage node.traverse (docutils <0.18) vs node.findall.
+    
+   TODO: This check can be removed when the minimum supported docutils version
+    for numpydoc is docutils>=0.18
+    """
     return (
         node.findall(condition, **kwargs)
         if hasattr(node, "findall")

--- a/numpydoc/numpydoc.py
+++ b/numpydoc/numpydoc.py
@@ -45,8 +45,8 @@ HASH_LEN = 12
 
 def _traverse_or_findall(node, condition, **kwargs):
     """Triage node.traverse (docutils <0.18) vs node.findall.
-    
-   TODO: This check can be removed when the minimum supported docutils version
+
+    TODO: This check can be removed when the minimum supported docutils version
     for numpydoc is docutils>=0.18
     """
     return (

--- a/numpydoc/numpydoc.py
+++ b/numpydoc/numpydoc.py
@@ -44,10 +44,10 @@ HASH_LEN = 12
 
 
 def _traverse_or_findall(node, condition, **kwargs):
-    """Triage node.traverse (docutils <0.18) vs node.findall.
+    """Triage node.traverse (docutils <0.18.1) vs node.findall.
 
     TODO: This check can be removed when the minimum supported docutils version
-    for numpydoc is docutils>=0.18
+    for numpydoc is docutils>=0.18.1
     """
     return (
         node.findall(condition, **kwargs)

--- a/numpydoc/tests/test_full.py
+++ b/numpydoc/tests/test_full.py
@@ -91,6 +91,8 @@ def test_reference(sphinx_app, html_file, expected_length):
     with open(op.join(out_dir, *html_file)) as fid:
         html = fid.read()
 
+    # TODO: This check can be removed when the minimum supported docutils version
+    # for numpydoc is docutils>=0.18
     pattern = (
         'role="doc-backlink"'
         if version.parse(docutils_version) >= version.parse("0.18")

--- a/numpydoc/tests/test_full.py
+++ b/numpydoc/tests/test_full.py
@@ -1,11 +1,13 @@
 import os.path as op
 import re
 import shutil
+from packaging import version
 
 import pytest
 import sphinx
 from sphinx.application import Sphinx
 from sphinx.util.docutils import docutils_namespace
+from docutils import __version__ as docutils_version
 
 
 # Test framework adapted from sphinx-gallery (BSD 3-clause)
@@ -89,7 +91,12 @@ def test_reference(sphinx_app, html_file, expected_length):
     with open(op.join(out_dir, *html_file)) as fid:
         html = fid.read()
 
-    reference_list = re.findall(r'<a role="doc-backlink" href="\#id\d+">(.*)<\/a>', html)
+    pattern = (
+        'role="doc-backlink"'
+        if version.parse(docutils_version) >= version.parse("0.18")
+        else 'class="fn-backref"'
+    )
+    reference_list = re.findall(rf'<a {pattern} href="\#id\d+">(.*)<\/a>', html)
 
     assert len(reference_list) == expected_length
     for ref in reference_list:

--- a/numpydoc/tests/test_full.py
+++ b/numpydoc/tests/test_full.py
@@ -89,7 +89,7 @@ def test_reference(sphinx_app, html_file, expected_length):
     with open(op.join(out_dir, *html_file)) as fid:
         html = fid.read()
 
-    reference_list = re.findall(r'<a class="fn-backref" href="\#id\d+">(.*)<\/a>', html)
+    reference_list = re.findall(r'<a role="doc-backlink" href="\#id\d+">(.*)<\/a>', html)
 
     assert len(reference_list) == expected_length
     for ref in reference_list:

